codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -844,57 +843,58 @@ public void bind(SocketAddress localAddress, ChannelPromise promise) {
 
             @Override
             public void disconnect(ChannelPromise promise) {
-                if (!promise.setUncancellable()) {
-                    return;
-                }
                 close(promise);
             }
 
             @Override
-            public void close(ChannelPromise promise) {
+            public void close(final ChannelPromise promise) {
                 if (!promise.setUncancellable()) {
                     return;
                 }
-                if (closePromise.isDone()) {
-                    promise.setFailure(new ClosedChannelException());
+                if (closeInitiated) {
+                    if (closePromise.isDone()) {
```

Comment:
From my reading of this method, the first time `close` is called the `closePromise` will be completed on line 897, so this should always evaluate to true. The major behavior change is that instead of completing the promise like `promise.setFailure(new ClosedChannelException())` we now do `promise.setSuccess()` which appears to be consistent with the behavior of the `AbstractUnsafe` implementation.

So, it seems like the state `closeInitiated` is redundant with the state of `closePromise` and as such the block
```java
             if (closeInitiated) {
                    if (closePromise.isDone()) {
                        // Closed already.
                        promise.setSuccess();
                    } else if (!(promise instanceof VoidChannelPromise)) { // Only needed if no VoidChannelPromise.
                        // This means close() was called before so we just register a listener and return
                        closePromise.addListener(new ChannelFutureListener() {
                            @Override
                            public void operationComplete(ChannelFuture future) throws Exception {
                                promise.setSuccess();
                            }
                        });
                    }
                    return;
                }
```
can be reduced to
```java
if (closePromise.isDone()) {
        // Closed already.
        promise.setSuccess(); // safe on VoidPromise
        return;
}
```

Reply:
@bryce-anderson I think the difference here is that if we do it with the listener is that it will be done in the "correct" order if close(...) is called in a re-entrance way.

Reply:
Ah, you're totally right. From looking at `AbstractUnsafe`, it appears that `closeFuture` is satisfied before the passed `promise` in the `doClose0` method, and that happens before calling the `fireChannelInactiveAndDeregister` method. If we switch the order of 
```java
promise.setSuccess();
closePromise.setSuccess();
```
to mirror that and place it before the inactive and deregister logic, we might avoid any reentrance issues outside of the write and flush of the RST frame, which also seems safe to move to after the satisfaction of `closePromise`.

Reply:
@bryce-anderson so you are happy with my patch ?

Reply:
I think this is definitely a step in the right direction, so I'm good with this patch. I think one of our root problems is that the `StreamChannel` will report `.isOpen() == true` even within a `channelInactive` event because `.isOpen()` is predicated on the state of the `closePromise` which gets completed as the very last step. It's not clear to me one way or the other if this is expected behavior, but I suspect it is not.

More specifically, we encounter reentrant behavior because we predicate a call to `Channel.close()` on `Channel.isClosed()` in response to a `channelInactive` event.
See string of events here:
https://github.com/twitter/finagle/blob/develop/finagle-netty4/src/main/scala/com/twitter/finagle/netty4/transport/ChannelTransport.scala#L188
https://github.com/twitter/finagle/blob/develop/finagle-netty4/src/main/scala/com/twitter/finagle/netty4/transport/ChannelTransport.scala#L94
https://github.com/twitter/finagle/blob/develop/finagle-netty4/src/main/scala/com/twitter/finagle/netty4/transport/ChannelTransport.scala#L141

It's entirely possible that we're misusing the event model.

Reply:
@bryce-anderson I think thats actually another issue... Let us fix the issue that was reported here first and then I will fix the other thing you brought up on another PR... WDYT ?

Reply:
That seems reasonable. Do you want to open the issue or should I?

Reply:
please do @bryce-anderson